File: android/app/src/main/java/com/airbridge/app/transport/connection/TlsMessageChannel.kt
package com.airbridge.app.transport.connection

import com.airbridge.app.transport.interfaces.IMessageChannel
import com.airbridge.app.transport.protocol.MessageType
import com.airbridge.app.transport.protocol.ProtocolMessage
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import android.util.Log
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.EOFException
import java.net.Socket
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import javax.net.ssl.SSLSocket

/**
 * [IMessageChannel] implementation over a TLS [SSLSocket].
 *
 * ## Wire format
 * Each frame on the wire is:
 * ```
 * ┌───────────────────────┬────────────┬───────────────────────┐
 * │  length  (4 bytes BE) │ type byte  │  payload (length)     │
 * └───────────────────────┴────────────┴───────────────────────┘
 * ```
 * `length` = payload size only (type byte is NOT included).
 *
 * The maximum payload is [ProtocolMessage.MAX_PAYLOAD_BYTES].
 *
 * ## Thread safety
 * [send] is protected by a [Mutex]; multiple coroutines may call it concurrently.
 *
 * ## Keepalive
 * PING (0xF0) and PONG (0xF1) frames are handled internally and never emitted on
 * [incomingMessages].  Call [startKeepalive] after the HANDSHAKE exchange to enable
 * the 30-second keepalive loop.  If no PONG is received within 10 seconds of a PING,
 * the channel is closed automatically.
 *
 * @param socket         The established TLS socket.
 * @param remoteDeviceId Stable device ID of the remote peer, resolved from the HANDSHAKE exchange
 *                       by [TlsConnectionManager] before the channel is constructed.
 */
class TlsMessageChannel(
    private val socket: Socket,
    override val remoteDeviceId: String,
    /** Override keepalive interval — useful in tests to avoid 30-second waits. */
    private val keepaliveIntervalMs: Long = KEEPALIVE_INTERVAL_MS,
    /** Override PONG timeout — useful in tests. */
    private val pongTimeoutMs: Long = PONG_TIMEOUT_MS
) : IMessageChannel {

    private val _connected = AtomicBoolean(true)
    override val isConnected: Boolean get() = _connected.get()

    /** Job for the active keepalive loop — cancelled in [close] so it stops immediately. */
    private var keepaliveJob: Job? = null

    private val out = DataOutputStream(BufferedOutputStream(socket.getOutputStream()))
    private val input  = DataInputStream(BufferedInputStream(socket.getInputStream()))

    private val sendMutex = Mutex()

    // Epoch-millisecond timestamp of the last PONG received from the peer.
    // Initialised to "now" so a freshly-created channel is not immediately considered dead.
    private val lastPongMs = AtomicLong(System.currentTimeMillis())

    companion object {
        private const val TAG                   = "AirBridge/Channel"
        private const val KEEPALIVE_INTERVAL_MS = 30_000L
        private const val PONG_TIMEOUT_MS       = 10_000L
    }

    // -------------------------------------------------------------------------
    // Incoming message stream
    // -------------------------------------------------------------------------

    /**
     * Cold [Flow] that reads framed messages until the connection closes cleanly.
     * Each subscription starts a new read loop — intended to be collected by a single consumer.
     * Completes normally on EOF; throws on protocol violations.
     *
     * PING and PONG messages are handled internally and never emitted to collectors.
     */
    override val incomingMessages: Flow<ProtocolMessage> = flow {
        var cancelledByCollector = false
        try {
            while (_connected.get()) {
                // Read 4-byte payload-size prefix (big-endian, does NOT include type byte)
                val payloadSize = try {
                    input.readInt()
                } catch (e: EOFException) {
                    Log.i(TAG, "[$remoteDeviceId] EOF — channel closed cleanly")
                    break  // clean close
                }

                if (payloadSize < 0) {
                    throw IllegalStateException("Invalid payload size: $payloadSize")
                }
                if (payloadSize > ProtocolMessage.MAX_PAYLOAD_BYTES) {
                    throw IllegalStateException(
                        "Payload too large: $payloadSize > ${ProtocolMessage.MAX_PAYLOAD_BYTES}"
                    )
                }

                // Read 1-byte message type
                val typeByte = input.readByte()
                val messageType = MessageType.fromByte(typeByte)

                // Read payload
                val payload = if (payloadSize > 0) {
                    ByteArray(payloadSize).also { input.readFully(it) }
                } else {
                    ByteArray(0)
                }

                // ── Keepalive intercept ───────────────────────────────────────
                // PING/PONG handled internally; never emitted to collectors.
                when (messageType) {
                    MessageType.PING -> {
                        try { send(ProtocolMessage(MessageType.PONG, ByteArray(0))) } catch (_: Exception) {}
                    }
                    MessageType.PONG -> {
                        lastPongMs.set(System.currentTimeMillis())
                    }
                    else -> {
                        Log.d(TAG, "[$remoteDeviceId] RX type=${messageType} len=${payload.size}")
                        emit(ProtocolMessage(type = messageType, payload = payload))
                    }
                }
            }
        } catch (e: CancellationException) {
            // Normal flow cancellation (e.g. first{} operator used by pairing service).
            // Do NOT mark the channel as disconnected — the socket is still alive.
            cancelledByCollector = true
            throw e
        } catch (e: Exception) {
            Log.e(TAG, "[$remoteDeviceId] Read loop error: ${e.javaClass.simpleName}: ${e.message}", e)
            _connected.set(false)
            throw e
        } finally {
            // For clean EOF exits and real errors, mark disconnected.
            // For collector cancellation (first{} etc.), leave connected so the next
            // collect() call can start a new read loop on the same socket.
            if (!cancelledByCollector) _connected.set(false)
            Log.i(TAG, "[$remoteDeviceId] incomingMessages flow ended, connected=${_connected.get()}")
        }
    }.flowOn(Dispatchers.IO)

    // -------------------------------------------------------------------------
    // Send
    // -------------------------------------------------------------------------

    /**
     * Serializes and sends [message] over the TLS socket.
     *
     * Protected by a [Mutex] so concurrent callers are serialized safely.
     * Dispatches all I/O to [Dispatchers.IO].
     */
    override suspend fun send(message: ProtocolMessage): Unit = withContext(Dispatchers.IO) {
        if (!_connected.get()) throw IllegalStateException("Channel is closed")
        val payload = message.payload
        if (payload.size > ProtocolMessage.MAX_PAYLOAD_BYTES) {
            throw IllegalArgumentException(
                "Payload too large: ${payload.size} > ${ProtocolMessage.MAX_PAYLOAD_BYTES}"
            )
        }

        sendMutex.withLock {
            // 4-byte header = payload size only (type byte is NOT included)
            out.writeInt(payload.size)
            out.writeByte(message.type.value.toInt())
            if (payload.isNotEmpty()) out.write(payload)
            out.flush()
        }
    }

    // -------------------------------------------------------------------------
    // Keepalive
    // -------------------------------------------------------------------------

    /**
     * Starts a background keepalive loop within [scope].
     * Sends a PING every 30 seconds; closes the channel if no PONG arrives within 10 seconds.
     * Should be called once, immediately after the HANDSHAKE exchange completes.
     */
    fun startKeepalive(scope: CoroutineScope) {
        keepaliveJob = scope.launch(Dispatchers.IO) {
            while (isActive && _connected.get()) {
                delay(keepaliveIntervalMs)
                if (!_connected.get()) break

                val pingTime = System.currentTimeMillis()
                val sendOk = try {
                    send(ProtocolMessage(MessageType.PING, ByteArray(0)))
                    Log.d(TAG, "[$remoteDeviceId] PING sent")
                    true
                } catch (e: Exception) {
                    Log.e(TAG, "[$remoteDeviceId] PING send failed: ${e.message}")
                    false
                }
                if (!sendOk) break

                // Wait for PONG (poll at 250ms intervals up to pongTimeoutMs)
                val deadline = pingTime + pongTimeoutMs
                while (System.currentTimeMillis() < deadline) {
                    if (lastPongMs.get() >= pingTime) break
                    delay(250)
                }

                if (lastPongMs.get() < pingTime) {
                    // No PONG received — treat as dead connection
                    Log.w(TAG, "[$remoteDeviceId] PONG timeout — closing channel")
                    close()
                    break
                } else {
                    Log.d(TAG, "[$remoteDeviceId] PONG received OK")
                }
            }
        }
    }

    // -------------------------------------------------------------------------
    // Close
    // -------------------------------------------------------------------------

    /** Closes the underlying socket, completing [incomingMessages]. */
    override suspend fun close(): Unit = withContext(Dispatchers.IO) {
        _connected.set(false)
        keepaliveJob?.cancel()   // stop the keepalive loop immediately, don't wait for the next delay()
        keepaliveJob = null
        runCatching { socket.close() }
    }
}

File: windows/AirBridge.Transport/Connection/TlsMessageChannel.cs
using System.Buffers.Binary;
using System.Net.Security;
using System.Net.Sockets;
using AirBridge.Transport.Interfaces;
using AirBridge.Transport.Protocol;

namespace AirBridge.Transport.Connection;

/// <summary>
/// <see cref="IMessageChannel"/> implementation that sends and receives
/// <see cref="ProtocolMessage"/> objects over an <see cref="SslStream"/> using
/// a simple length-prefix wire format:
/// <code>
/// [4 bytes big-endian payload length][1 byte MessageType][N bytes payload]
/// </code>
/// <para>
/// The length field encodes the number of payload bytes only (not the type byte).
/// </para>
/// <para>Sends are serialised with a <see cref="SemaphoreSlim"/> so callers may
/// call <see cref="SendAsync"/> concurrently from multiple threads.</para>
/// <para>
/// PING (0xF0) / PONG (0xF1) keepalive frames are handled internally and never
/// surfaced to callers via <see cref="ReceiveAsync"/>.  Call
/// <see cref="StartKeepaliveAsync"/> after construction to enable the keepalive loop.
/// </para>
/// </summary>
public sealed class TlsMessageChannel : IMessageChannel
{
    private readonly SslStream _stream;
    private readonly SemaphoreSlim _writeLock = new(1, 1);
    private volatile bool _connected = true;
    private bool _disposed;

    // ── Keepalive state ────────────────────────────────────────────────────
    private static readonly TimeSpan DefaultKeepaliveInterval = TimeSpan.FromSeconds(30);
    private static readonly TimeSpan DefaultPongTimeout       = TimeSpan.FromSeconds(10);

    private readonly TimeSpan _keepaliveInterval;
    private readonly TimeSpan _pongTimeout;

    /// <summary>
    /// Linked to the caller's token AND cancelled in <see cref="DisposeAsync"/>.
    /// Ensures the keepalive loop stops immediately when the channel is disposed,
    /// even if the caller's token is long-lived (e.g. the listener's lifetime token).
    /// </summary>
    private CancellationTokenSource? _keepaliveCts;

    /// <summary>
    /// UTC timestamp of the last PONG received from the peer.
    /// Updated by the receive loop whenever a PONG arrives.
    /// </summary>
    private DateTime _lastPongUtc = DateTime.UtcNow;

    // ── IMessageChannel ────────────────────────────────────────────────────

    /// <summary>
    /// Identifier of the remote peer.  Set to the peer's IP address initially;
    /// updated to the peer's stable device ID after the HANDSHAKE exchange.
    /// </summary>
    public string RemoteDeviceId { get; internal set; }

    /// <inheritdoc/>
    public bool IsConnected => _connected;

    /// <inheritdoc/>
    public event EventHandler? Disconnected;

    /// <summary>
    /// Wraps an already-authenticated <see cref="SslStream"/> as a message channel.
    /// </summary>
    /// <param name="stream">An authenticated, open SSL stream.</param>
    /// <param name="remoteDeviceId">
    /// Identifier of the remote peer (filled from the Handshake message in higher layers;
    /// may be a placeholder until the handshake completes).
    /// </param>
    /// <param name="keepaliveInterval">How often to send a PING. Defaults to 30 s.</param>
    /// <param name="pongTimeout">How long to wait for a PONG before closing. Defaults to 10 s.</param>
    public TlsMessageChannel(
        SslStream stream,
        string remoteDeviceId,
        TimeSpan? keepaliveInterval = null,
        TimeSpan? pongTimeout = null)
    {
        _stream            = stream        ?? throw new ArgumentNullException(nameof(stream));
        RemoteDeviceId     = remoteDeviceId ?? throw new ArgumentNullException(nameof(remoteDeviceId));
        _keepaliveInterval = keepaliveInterval ?? DefaultKeepaliveInterval;
        _pongTimeout       = pongTimeout       ?? DefaultPongTimeout;
    }

    // ── IMessageChannel ────────────────────────────────────────────────────

    /// <inheritdoc/>
    /// <remarks>
    /// Wire format: <c>[4-byte big-endian payload length][1-byte type][payload]</c>.
    /// Thread-safe via an internal semaphore.
    /// </remarks>
    public async Task SendAsync(ProtocolMessage message, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(message);
        if (message.Payload.Length > ProtocolMessage.MaxPayloadBytes)
            throw new ArgumentException(
                $"Payload length {message.Payload.Length} exceeds MaxPayloadBytes ({ProtocolMessage.MaxPayloadBytes}).",
                nameof(message));

        // Frame: [4-byte length][1-byte type][payload]
        var frame = new byte[4 + 1 + message.Payload.Length];
        BinaryPrimitives.WriteUInt32BigEndian(frame, (uint)message.Payload.Length);
        frame[4] = (byte)message.Type;
        message.Payload.CopyTo(frame, 5);

        // Guard: if the channel has already been disposed (e.g. session replacement
        // disposed the old channel while a file-transfer chunk was mid-flight), fail
        // fast instead of letting _writeLock.WaitAsync throw ObjectDisposedException
        // from the SemaphoreSlim internals, which would surface as an unhandled exception.
        if (_disposed)
            throw new ObjectDisposedException(nameof(TlsMessageChannel), $"Cannot send on a disposed channel (remote={RemoteDeviceId}).");

        bool lockAcquired = false;
        try
        {
            await _writeLock.WaitAsync(cancellationToken).ConfigureAwait(false);
            lockAcquired = true;
            await _stream.WriteAsync(frame, cancellationToken).ConfigureAwait(false);
            await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex) when (IsNetworkException(ex))
        {
            AirBridge.Core.AppLog.Warn($"[{RemoteDeviceId}] Send failed (type={message.Type}) — {ex.GetType().Name}: {ex.Message}");
            SignalDisconnect();
            throw;
        }
        finally
        {
            if (lockAcquired && !_disposed) _writeLock.Release();
        }
    }

    /// <inheritdoc/>
    /// <returns>
    /// The next <see cref="ProtocolMessage"/>, or <c>null</c> if the peer closed
    /// the connection cleanly (zero bytes read on the header).
    /// PING and PONG messages are handled internally and are never returned to callers.
    /// </returns>
    /// <exception cref="InvalidDataException">
    /// Thrown when the payload length in the frame header exceeds
    /// <see cref="ProtocolMessage.MaxPayloadBytes"/>.
    /// </exception>
    public async Task<ProtocolMessage?> ReceiveAsync(CancellationToken cancellationToken = default)
    {
        while (true)
        {
            // ── Read 4-byte length header ──────────────────────────────────────
            var header = new byte[4];
            int bytesRead;
            try
            {
                bytesRead = await ReadExactAsync(_stream, header, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception ex) when (IsNetworkException(ex))
            {
                // On Windows, SslStream throws a SocketException (connection reset /
                // forcibly closed) when the remote side closes without a TLS close_notify.
                // Treat any network exception during the header read as a clean disconnect
                // and return null — the same as a graceful 0-byte EOF.
                AirBridge.Core.AppLog.Warn($"[{RemoteDeviceId}] Silent socket death detected during header read — {ex.GetType().Name}: {ex.Message}");
                SignalDisconnect();
                return null;
            }

            if (bytesRead == 0)
            {
                // Clean close from remote side (graceful TLS close_notify)
                AirBridge.Core.AppLog.Info($"[{RemoteDeviceId}] Clean EOF — peer closed TLS connection gracefully");
                _connected = false;
                return null;
            }

            uint payloadLength = BinaryPrimitives.ReadUInt32BigEndian(header);

            if (payloadLength > ProtocolMessage.MaxPayloadBytes)
                throw new InvalidDataException(
                    $"Incoming payload length {payloadLength} exceeds MaxPayloadBytes ({ProtocolMessage.MaxPayloadBytes}).");

            // ── Read 1-byte message type + payload ────────────────────────────
            var body = new byte[1 + payloadLength];
            try
            {
                bytesRead = await ReadExactAsync(_stream, body, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception ex) when (IsNetworkException(ex))
            {
                AirBridge.Core.AppLog.Warn($"[{RemoteDeviceId}] Silent socket death detected during body read (payloadLen={payloadLength}) — {ex.GetType().Name}: {ex.Message}");
                SignalDisconnect();
                throw;
            }

            if (bytesRead == 0)
            {
                AirBridge.Core.AppLog.Info($"[{RemoteDeviceId}] Unexpected EOF mid-frame (payloadLen={payloadLength})");
                _connected = false;
                return null;
            }

            var type    = (MessageType)body[0];
            var payload = body[1..];

            // ── Keepalive intercept ───────────────────────────────────────────
            if (type == MessageType.Ping)
            {
                // Reply with PONG immediately; do not surface PING to caller
                try
                {
                    await SendAsync(
                        new ProtocolMessage(MessageType.Pong, Array.Empty<byte>()),
                        cancellationToken).ConfigureAwait(false);
                }
                catch { /* best-effort; if send fails the loop will catch it on next read */ }
                continue; // read next message
            }

            if (type == MessageType.Pong)
            {
                _lastPongUtc = DateTime.UtcNow;
                continue; // do not surface PONG to caller
            }

            return new ProtocolMessage(type, payload);
        }
    }

    // ── Keepalive ──────────────────────────────────────────────────────────

    /// <summary>
    /// Starts a background keepalive loop that sends a PING every 30 seconds and
    /// closes the channel if no PONG arrives within 10 seconds.
    /// Should be called once, immediately after the HANDSHAKE exchange completes.
    /// </summary>
    /// <param name="cancellationToken">
    /// Token that, when cancelled, gracefully terminates the keepalive loop.
    /// </param>
    public Task StartKeepaliveAsync(CancellationToken cancellationToken)
    {
        // Create an internal CTS linked to the caller's token so we can independently
        // cancel the keepalive loop from DisposeAsync, regardless of how long-lived
        // the caller's token is.
        _keepaliveCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        return Task.Run(() => KeepaliveLoopAsync(_keepaliveCts.Token), _keepaliveCts.Token);
    }

    private async Task KeepaliveLoopAsync(CancellationToken cancellationToken)
    {
        try
        {
            while (!cancellationToken.IsCancellationRequested && _connected)
            {
                await Task.Delay(_keepaliveInterval, cancellationToken).ConfigureAwait(false);

                if (!_connected) break;

                var pingTime = DateTime.UtcNow;
                AirBridge.Core.AppLog.Info($"[{RemoteDeviceId}] PING sent");
                try
                {
                    await SendAsync(
                        new ProtocolMessage(MessageType.Ping, Array.Empty<byte>()),
                        cancellationToken).ConfigureAwait(false);
                }
                catch
                {
                    // Send failure already signals disconnect; stop the loop
                    break;
                }

                // Wait for PONG (poll at 250ms intervals up to _pongTimeout)
                var deadline = pingTime + _pongTimeout;
                while (DateTime.UtcNow < deadline)
                {
                    if (_lastPongUtc >= pingTime) break;
                    await Task.Delay(250, cancellationToken).ConfigureAwait(false);
                }

                if (_lastPongUtc < pingTime)
                {
                    // No PONG received within timeout — treat as dead connection
                    AirBridge.Core.AppLog.Warn($"[{RemoteDeviceId}] PONG timeout — closing channel");
                    SignalDisconnect();
                    await DisposeAsync().ConfigureAwait(false);
                    break;
                }

                AirBridge.Core.AppLog.Info($"[{RemoteDeviceId}] PONG received OK");
            }
        }
        catch (OperationCanceledException) { /* normal shutdown */ }
    }

    // ── IAsyncDisposable ───────────────────────────────────────────────────

    /// <inheritdoc/>
    public async ValueTask DisposeAsync()
    {
        if (_disposed) return;
        _disposed  = true;
        _connected = false;
        // Cancel the keepalive loop immediately so it does not linger after the channel
        // is disposed.  Without this, the loop sits in Task.Delay(30 s) even after
        // _connected=false, causing ghost PINGs on dead channels minutes later.
        _keepaliveCts?.Cancel();
        _keepaliveCts?.Dispose();
        _keepaliveCts = null;
        await _stream.DisposeAsync().ConfigureAwait(false);
        _writeLock.Dispose();
    }

    // ── Private helpers ────────────────────────────────────────────────────

    /// <summary>
    /// Reads exactly <paramref name="buffer"/>.Length bytes from <paramref name="stream"/>.
    /// Returns 0 if the very first read returns 0 (clean EOF before any bytes).
    /// Throws <see cref="EndOfStreamException"/> if the stream closes after some bytes
    /// have been read but before the buffer is full (mid-frame disconnect).
    /// </summary>
    private static async Task<int> ReadExactAsync(
        Stream stream,
        byte[] buffer,
        CancellationToken cancellationToken)
    {
        int offset = 0;
        while (offset < buffer.Length)
        {
            int n = await stream.ReadAsync(buffer.AsMemory(offset, buffer.Length - offset), cancellationToken)
                                .ConfigureAwait(false);
            if (n == 0)
            {
                if (offset == 0)
                    return 0; // clean EOF — no bytes read yet
                // Mid-frame EOF: connection dropped after partial data
                throw new EndOfStreamException(
                    $"Connection closed after {offset} of {buffer.Length} expected bytes.");
            }
            offset += n;
        }
        return offset;
    }

    private void SignalDisconnect()
    {
        if (_connected)
        {
            _connected = false;
            Disconnected?.Invoke(this, EventArgs.Empty);
        }
    }

    private static bool IsNetworkException(Exception ex) =>
        ex is IOException or SocketException or ObjectDisposedException or EndOfStreamException;
}
